src/neo4j_genai/components/entity_relation_extractor.py

```diff
@@ -0,0 +1,114 @@
+
+from neo4j_genai.components.types import Neo4jGraph, TextChunks
+from neo4j_genai.exceptions import LLMGenerationError
+from neo4j_genai.generation.prompts import ERExtractionTemplate, PromptTemplate
+from neo4j_genai.llm import LLMInterface
+from neo4j_genai.pipeline.component import Component
+
+logger = logging.getLogger(__name__)
+
+
+class EntityRelationExtractor(Component):
+    async def run(self, chunks: TextChunks, **kwargs: Any) -> Neo4jGraph:
+        # for each chunk, returns a dict with entities and relations keys
+        return Neo4jGraph(nodes=[], relationships=[])
+
+
+class OnError(enum.Enum):
+    RAISE = "RAISE"
+    IGNORE = "CONTINUE"
+
+
+class LLMEntityRelationExtractor(EntityRelationExtractor):
+    def __init__(
+        self,
+        llm: LLMInterface,
+        prompt_template: ERExtractionTemplate | str = ERExtractionTemplate(),
+        on_error: OnError = OnError.RAISE,
+    ) -> None:
+        self.llm = llm  # with response_format={ "type": "json_object" },
+        if isinstance(prompt_template, str):
+            template = PromptTemplate(prompt_template, expected_inputs=[])
+        else:
+            template = prompt_template
+        self.prompt_template = template
+        self.on_error = on_error
+
+    def update_ids(self, chunk_index: int, graph: Neo4jGraph) -> Neo4jGraph:
+        for node in graph.nodes:
+            node.id = f"{chunk_index}:{node.id}"
+            if node.properties is None:
+                node.properties = {}
+            node.properties.update({"chunk_index": chunk_index})
+        for rel in graph.relationships:
+            rel.start_node_id = f"{chunk_index}:{rel.start_node_id}"
+            rel.end_node_id = f"{chunk_index}:{rel.end_node_id}"
+        return graph
+
+    @validate_call
+    async def run(
+        self,
+        chunks: TextChunks,
+        schema: Union[BaseModel, Dict[str, Any], None] = None,
+        examples: str = "",
+        **kwargs: Any,
+    ) -> Neo4jGraph:
+        schema = schema or {}
+        examples = examples or ""
+        graph = Neo4jGraph()
+        for chunk_index, chunk in enumerate(chunks.chunks):
+            prompt = self.prompt_template.format(
+                text=chunk.text, schema=schema, examples=examples
+            )
+            llm_result = self.llm.invoke(prompt)
+            try:
+                result = json.loads(llm_result.content)
+            except json.JSONDecodeError:
+                if self.on_error == OnError.RAISE:
+                    raise LLMGenerationError(
+                        f"LLM response is not valid JSON {llm_result.content}"
+                    )
+                else:
+                    logger.error(
+                        f"LLM response is not valid JSON {llm_result.content} for chunk_index={chunk_index}"
+                    )
+                result = {"nodes": [], "relationships": []}
+            try:
+                chunk_graph = Neo4jGraph(**result)
```

Comment:
How do we connect the chunks to the extracted nodes using this approach? Do we not need to do it manually after the graph has been extracted? I'm also wondering if we might want to point the chunks to each other i.e. include a `NEXT_CHUNK` relationship which links them?

Reply:
Entities have a "chunk_index" property, so that we can save them. But at the moment, chunks are not included as nodes in the graph. Do you think it should be done in this component? I'm not sure we're still in the scope of the entity and relation extraction.

Reply:
I think Oskar mentioned we also want to be able to create lexical graphs (if that's the right terminology) which include the chunks and embeddings. If we do it would make sense to connect them to the other chunks and associated nodes so the queries for retrievers like the `VectorCypherRetriever` are easy to write. We could include another component to connect the chunks to the rest of the graph but imo it also makes sense to do it here as we're already looping through the chunks so it wouldn't be that hard to create the `NEXT_CHUNK` relationship between the chunks. Then for each chunk we could create a `chunk` node and loop through the other nodes and create a relationship which joins them.

Reply:
Done in https://github.com/neo4j/neo4j-genai-python/pull/85/commits/0c46445ca58008eda71a727af8a57df8c68e3bb2

I think we'll see in the future if we need to have a separate component, could be useful for people who might want to first do the chunk extraction, save them, and then run ER extraction for instance, but we're not there yet.